src/PowerShellEditorServices.Engine/Services/TextDocument/Handlers/CompletionHandler.cs

```diff
@@ -0,0 +1,320 @@
+
+namespace Microsoft.PowerShell.EditorServices.TextDocument
+{
+    internal class CompletionHandler : ICompletionHandler, ICompletionResolveHandler
+    {
+        const int DefaultWaitTimeoutMilliseconds = 5000;
+        private readonly CompletionItem[] s_emptyCompletionResult = new CompletionItem[0];
+
+        private readonly ILogger _logger;
+        private readonly PowerShellContextService _powerShellContextService;
+        private readonly WorkspaceService _workspaceService;
+
+        private CompletionResults _mostRecentCompletions;
+
+        private int _mostRecentRequestLine;
+
+        private int _mostRecentRequestOffest;
+
+        private string _mostRecentRequestFile;
+
+        private CompletionCapability _capability;
+
+        public CompletionHandler(
+            ILoggerFactory factory,
+            PowerShellContextService powerShellContextService,
+            WorkspaceService workspaceService)
+        {
+            _logger = factory.CreateLogger<CompletionHandler>();
+            _powerShellContextService = powerShellContextService;
+            _workspaceService = workspaceService;
+        }
+
+        public CompletionRegistrationOptions GetRegistrationOptions()
+        {
+            return new CompletionRegistrationOptions
+            {
+                DocumentSelector = new DocumentSelector(new DocumentFilter { Pattern = "**/*.ps*1" }),
+                ResolveProvider = true,
+                TriggerCharacters = new[] { ".", "-", ":", "\\" }
+            };
+        }
+
+        public async Task<CompletionList> Handle(CompletionParams request, CancellationToken cancellationToken)
+        {
+            int cursorLine = (int) request.Position.Line + 1;
+            int cursorColumn = (int) request.Position.Character + 1;
+
+            ScriptFile scriptFile =
+                _workspaceService.GetFile(
+                    request.TextDocument.Uri.ToString());
+
+            CompletionResults completionResults =
+                await GetCompletionsInFileAsync(
+                    scriptFile,
+                    cursorLine,
+                    cursorColumn);
+
+            CompletionItem[] completionItems = s_emptyCompletionResult;
+
+            if (completionResults != null)
+            {
+                completionItems = new CompletionItem[completionResults.Completions.Length];
+                for (int i = 0; i < completionItems.Length; i++)
+                {
+                    completionItems[i] = CreateCompletionItem(completionResults.Completions[i], completionResults.ReplacedRange, i + 1);
+                }
+            }
+
+            return new CompletionList(completionItems);
+        }
+
+        public bool CanResolve(CompletionItem value)
+        {
+            return value.Kind == CompletionItemKind.Function;
+        }
+
+        public async Task<CompletionItem> Handle(CompletionItem request, CancellationToken cancellationToken)
```

Comment:
When is this called? Just on select? Or for every completion item?

Reply:
It's called when a completion item is highlighted in the completion list